netket/hilbert/discrete_hilbert.py

```diff
@@ -0,0 +1,203 @@
+# Copyright 2021 The NetKet Authors - All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import List, Tuple, Optional, Union, Iterator
+
+import numpy as np
+
+from .abstract_hilbert import AbstractHilbert
+
+max_states = np.iinfo(np.int32).max
+"""int: Maximum number of states that can be indexed"""
+
+
+class NoneTypeT:
+    pass
+
+
+NoneType = NoneTypeT()
+
+
+legacy_warn_str = (
+    "This choice of arguments for `hilbert.random_state` is deprecated and "
+    "will be removed in a future version.\n"
+    "The new syntax is\n"
+    "    hilbert.random_state(prngkey, size=1, dtype=jnp.float32)\n"
+    "and, like in `jax.random`, the first argument is mandatory and must be a valid jax `PRNGKey`."
+    "Results may differ between the states generated by the legacy, "
+    "deprecated code and the new one. Note that the old code does not "
+    "support defining custom methods."
+)
+
+
+class DiscreteHilbert(AbstractHilbert):
+    """Abstract class for an hilbert space defined on a lattice.
+
+    This class definese the common interface that can be used to
+    interact with hilbert spaces on lattices.
+    """
+
+    def __init__(self, shape: Tuple[int, ...]):
+        """
+        Initializes a discrete Hilbert space with a basis of given shape.
+
+        Args:
+            shape: The local dimension of the Hilbert space for each degree
+                of freedom.
+        """
+        self._shape = shape
+
+        super().__init__()
+
+    @property
+    def shape(self) -> Tuple[int, ...]:
+        r"""The size of the hilbert space on every site."""
+        return self._shape
+
+    @property
+    def is_discrete(self) -> bool:
```

Comment:
should we leave this method? I guess we could just check if something is derived from DiscreteHilbert ?

Reply:
I'd say we should deprecate or remove it.